File: scripts/predict_now.py
import pandas as pd
import numpy as np
import os
import joblib
import sys
from tensorflow.keras.models import load_model

# Paths
DATA_CSV = "Data/lstm_dataset.csv"
MODEL_PATH = "Data/lstm_model.h5"
SCALER_X_PATH = "Data/scaler_x.pkl"
SCALER_Y_PATH = "Data/scaler_y.pkl"

if not os.path.exists(MODEL_PATH) or not os.path.exists(SCALER_X_PATH):
    print("Error: The model or scalers do not exist. Please run train_lstm.py first.")
    sys.exit(1)

print("Loading model and scalers...")
model = load_model(MODEL_PATH)
scaler_x = joblib.load(SCALER_X_PATH)
scaler_y = joblib.load(SCALER_Y_PATH)

print(f"Loading latest market data from {DATA_CSV}...")
df = pd.read_csv(DATA_CSV)

feature_cols = [
    'supply_bottom_pips_60m', 'demand_top_pips_60m', 'demand_bottom_pips_60m',
    'trendline_upper_pips_60m', 'trendline_lower_pips_60m', 'channel_width_pips_60m',
    'ghost_high_pips_60m', 'ghost_low_pips_60m',
    'ob_top_pips', 'ob_bottom_pips', 'fvg_top_pips', 'fvg_bottom_pips',
    'bos_choch_pips', 'eq_level_pips', 'vwap_pips',
    'poc_pips', 'vah_pips', 'val_pips', 'atr_1m'
]

SEQ_LEN = 10

if len(df) < SEQ_LEN:
    print(f"Error: Not enough data for prediction. Need at least {SEQ_LEN} candles.")
    sys.exit(1)

# Get the last SEQ_LEN rows (current state)
df_current = df.tail(SEQ_LEN).copy()
df_current[feature_cols] = df_current[feature_cols].fillna(0)

# Scale
X_current_scaled = scaler_x.transform(df_current[feature_cols])
# Reshape for LSTM (1, SEQ_LEN, NUM_FEATURES)
X_input = X_current_scaled.reshape(1, SEQ_LEN, len(feature_cols))

print("Predicting target_5m (Price displacement in 5 minutes)...")
pred_scaled = model.predict(X_input)
pred = scaler_y.inverse_transform(pred_scaled)

target_value = pred[0][0]
current_close = df_current['close'].iloc[-1]
current_time = df_current['time'].iloc[-1]

print("\n=========================================")
print(f"CURRENT STATE (Time: {current_time})")
print(f"Current Close Price: {current_close:.2f}")
print("=========================================")
print(f"PREDICTION (5 minutes ahead):")
if target_value > 0:
    print(f"Direction: UP ⬆️")
    print(f"Expected Move: +{target_value:.2f} points")
    print(f"Projected Price: {current_close + target_value:.2f}")
else:
    print(f"Direction: DOWN ⬇️")
    print(f"Expected Move: {target_value:.2f} points")
    print(f"Projected Price: {current_close + target_value:.2f}")
print("=========================================\n")

File: scripts/predict_now.pl
#!/usr/bin/perl
use strict;
use warnings;
use FindBin;
use Storable;

my $data_csv = "$FindBin::Bin/../Data/lstm_dataset.csv";
my $model_path = "$FindBin::Bin/../Data/knn_model.dat";

if (!-e $model_path) {
    die "Error: The model does not exist. Please run train_model.pl first.\n";
}

print "Loading model and scalers...\n";
my $model = retrieve($model_path) or die "Cannot load model from $model_path!\n";
my $min = $model->{min};
my $max = $model->{max};
my $training_data = $model->{data};
my @feature_cols = @{$model->{features}};

print "Loading latest market data from $data_csv...\n";
open(my $fh, '<', $data_csv) or die "Error: $!\n";
my $header = <$fh>;
chomp $header;
my @cols = split(/,/, $header);
my %col_idx;
for my $i (0 .. $#cols) {
    $col_idx{$cols[$i]} = $i;
}

my @rows;
while (my $line = <$fh>) {
    chomp $line;
    push @rows, [split(/,/, $line)];
}
close($fh);

if (@rows == 0) {
    die "Error: Not enough data for prediction.\n";
}

my $last_row = $rows[-1];
my %current_features;
for my $f (@feature_cols) {
    my $val = $last_row->[$col_idx{$f}];
    $val = 0 if (!defined $val || $val eq '');
    $val += 0;
    
    # Scale feature
    my $scaled = ($val - $min->{$f}) / ($max->{$f} - $min->{$f});
    $current_features{$f} = $scaled;
}

print "Predicting target_5m using K-Nearest Neighbors (K=5)...\n";

# Calculate distances
my @distances;
for my $train_row (@$training_data) {
    my $sum_sq = 0;
    for my $f (@feature_cols) {
        my $train_val = $train_row->{$f};
        my $scaled_train = ($train_val - $min->{$f}) / ($max->{$f} - $min->{$f});
        my $diff = $current_features{$f} - $scaled_train;
        $sum_sq += $diff * $diff;
    }
    push @distances, {
        distance => sqrt($sum_sq),
        target => $train_row->{target_5m}
    };
}

# Sort by distance
@distances = sort { $a->{distance} <=> $b->{distance} } @distances;

# Take top 5
my $k = 5;
my $sum_target = 0;
for my $i (0 .. $k-1) {
    last if $i >= @distances;
    $sum_target += $distances[$i]->{target};
}
my $target_value = $sum_target / $k;

my $current_close = $last_row->[$col_idx{'close'}] + 0;
my $current_time = $last_row->[$col_idx{'time'}];

print "\n=========================================\n";
print "CURRENT STATE (Time: $current_time)\n";
printf "Current Close Price: %.2f\n", $current_close;
print "=========================================\n";
print "PREDICTION (5 minutes ahead):\n";
if ($target_value > 0) {
    print "Direction: UP \xE2\xAC\x86\xEF\xB8\x8F\n";
    printf "Expected Move: +%.2f points\n", $target_value;
    printf "Projected Price: %.2f\n", $current_close + $target_value;
} else {
    print "Direction: DOWN \xE2\xAC\x87\xEF\xB8\x8F\n";
    printf "Expected Move: %.2f points\n", $target_value;
    printf "Projected Price: %.2f\n", $current_close + $target_value;
}
print "=========================================\n\n";
